tensorflow/cc/training/optimizer.cc

```diff
@@ -0,0 +1,177 @@
+    // as the number of vars it means that at least
+    // one var from var_list is not a variable
+    if (var_list.size() != grads_and_vars->size()) {
+        scope.UpdateStatus(Status(::tensorflow::error::Code::INVALID_ARGUMENT,
+            "You are trying to compute the gradients of non Variable Output."));
+    }
+}
+
+std::vector<Output> Optimizer::ApplyGradients(const Scope& scope,
+                                              const GradAndVar& grads_and_vars) {
+    // we'll return a list of update ops
+    // TODO(theflofly): make a Group op and return one output
+    std::vector<Output> update_ops;
+
+    if (grads_and_vars.empty()) {
+        scope.UpdateStatus(Status(::tensorflow::error::Code::INVALID_ARGUMENT,
+                                  "No variable to update provided."));
+    }
+
+    Scope scope_optimizer = scope.NewSubScope(this->name_);
+
+    // each instance can prepare itself using their own logic
+    this->Prepare(scope_optimizer);
+
+    // retrieve the processor for each (gradient, var) tuple
+    for (int i = 0; i < grads_and_vars.size(); i++) {
+        Output grad = std::get<0>(grads_and_vars[i]);
+        Output var = std::get<1>(grads_and_vars[i]);
+        OptimizableVariable *processor = GetProcessor(scope_optimizer, var);
+
+        // no processor for this kind of variable, the scope.status()
+        // contains more details
+        if (processor == nullptr) {
+            return update_ops;
+        }
+
+        update_ops.push_back(processor->UpdateOp(scope_optimizer, *this, grad));
+        delete processor;
+    }
+
+    return update_ops;
+}
+
+void Optimizer::ComputeGradients(const Scope& scope,
+                                 const std::vector<Output>& loss,
+                                 const std::vector<Output>& var_list,
+                                 GradAndVar* grads_and_vars) {
+    Scope scope_gradient = scope.NewSubScope(kGradientsName);
+
+    // fill grad_loss with 'OnesLike' for all shapes in 'loss'
+    std::vector<Output> grad_loss;
+    grad_loss.reserve(loss.size());
+
+    for (const Output& loss_output : loss) {
+        grad_loss.emplace_back(ops::OnesLike(scope_gradient, loss_output));
+    }
+
+    ComputeGradients(scope_gradient, loss, var_list, grad_loss, grads_and_vars);
+}
+
+std::vector<Output> Optimizer::Minimize(const Scope& scope,
+                                        const std::vector<Output>& loss,
+                                        const std::vector<Output>& var_list) {
+    GradAndVar grad_and_var;
+    ComputeGradients(scope, loss, var_list, &grad_and_var);
+    return ApplyGradients(scope, grad_and_var);
+}
+
+std::vector<Output> Optimizer::Minimize(const Scope& scope,
+                                        const std::vector<Output>& loss) {
+    // retrieve all the variables from the graph
+    std::vector<Output> var_list;
+
+    for (Node* node : scope.graph()->nodes()) {
+        if (::tensorflow::grappler::IsVariable(node->def())) {
```

Comment:
Do you know how I can do that ? Because we don't have a collection of trainable vars as opposed to the python version where we use `ops.GraphKeys.TRAINABLE_VARIABLES` ?